algebird-core/src/main/scala/com/twitter/algebird/CMSHasher.scala

```diff
@@ -0,0 +1,150 @@
+    }
+
+  }
+
+  implicit val cmsHasherShort: CMSHasher[Short] = CMSHasherInt.contramap(x => x.toInt)
+
+  implicit object CMSHasherInt extends CMSHasher[Int] {
+
+    override def hash(a: Int, b: Int, width: Int)(x: Int): Int = {
+      val unModded: Int = (x * a) + b
+      val modded: Long = (unModded + (unModded >> 32)) & Int.MaxValue
+      modded.toInt % width
+    }
+  }
+
+  /**
+   * =Implementation details=
+   *
+   * This hash function is based upon Murmur3.  Note that the original CMS paper requires
+   * `d` (depth) pair-wise independent hash functions;  in the specific case of Murmur3 we argue that it is sufficient
+   * to pass `d` different seed values to Murmur3 to achieve a similar effect.
+   *
+   * To seed Murmur3 we use only `a`, which is a randomly drawn `Int` via [[scala.util.Random]] in the CMS code.
+   * What is important to note is that we intentionally ignore `b`.  Why?  We need to ensure that we seed Murmur3 with
+   * a random value, notably one that is uniformly distributed.  Somewhat surprisingly, combining two random values
+   * (such as `a` and `b` in our case) typically worsens the "randomness" of the combination, i.e. the combination is
+   * less uniformly distributed as either of its original inputs.  Hence the combination of two random values is
+   * discouraged in this context, notably if the two random inputs were generated from the same source anyways, which
+   * is the case for us because we use Scala's PRNG only.
+   *
+   * For further details please refer to the discussion
+   * [[http://stackoverflow.com/questions/3956478/understanding-randomness Understanding Randomness]] on
+   * StackOverflow.
+   *
+   * @param a Must be a random value, typically created via [[scala.util.Random]].
+   * @param b Ignored by this particular hash function, see the reasoning above for the justification.
+   * @param width Width of the CMS counting table, i.e. the width/size of each row in the counting table.
+   * @param x Item to be hashed.
+   * @return Slot assigned to item `x` in the vector of size `width`, where `x in [0, width)`.
+   */
+  private[algebird] def hashBytes(a: Int, b: Int, width: Int)(x: Array[Byte]): Int = {
+    val hash: Int = scala.util.hashing.MurmurHash3.arrayHash(x, a)
+    // We only want positive integers for the subsequent modulo.  This method mimics Java's Hashtable
+    // implementation.  The Java code uses `0x7FFFFFFF` for the bit-wise AND, which is equal to Int.MaxValue.
+    val positiveHash = hash & Int.MaxValue
+    positiveHash % width
+  }
+
+  // This CMSHasher[BigInt] is newer, and faster, than the old version
+  // found in CMSHasherImplicits. Unless you have serialized data that
+  // requires the old implementation for correctness, you should be
+  // using this instance.
+  implicit object CMSHasherBigInt extends CMSHasher[BigInt] {
+    override def hash(a: Int, b: Int, width: Int)(x: BigInt): Int =
+      ((a * x.hashCode + b) & Int.MaxValue) % width
```

Comment:
actually, one concern I have about this hashCode is that if you collide on the `Int` hashCode, you are already sunk, no more mixing will help. You are not going to be using `BigInt` unless you really have big numbers (bigger than Long) for instance, but actually this hash will have a lot more collisions than `Long`.

Since `BigInt` hashing is not super common, I'm not sure we should make this a default when we have a concern about loss of precision.

An alternate approach might be to: `(a * x + b).hashCode & Int.MaxValue` so that we don't hash until we have already done the linear transform. This gives two different a,b values the chance to not collide. I wonder how much slower that will be.